prime-router/docs/design/design/deduplication/deduplication.md

```diff
@@ -0,0 +1,248 @@
+    )
+}
+```
+
+### Deduplication Workflow Placement
+
+<BR><img alt="Dedupe_Design.jpg" height="310" src="Dedupe_Design.jpg" width="450"/>
+
+- The deduplication workflow will happen in the middle of the Convert Step. After conversion related processing, but before any of the queue related processing. ([FHIRConverter.kt#L541](https://github.com/CDCgov/prime-reportstream/blob/c942a9a6f6be347d82196939e1cf677512af4a06/prime-router/src/main/kotlin/fhirengine/engine/FHIRConverter.kt#L541)).
+  - Before the Deduplication Workflow, the Convert Step will:
+    - Validate the items
+    - Convert to FHIR, if necessary
+    - De-batch items from the report
+    - Convert these items into FHIR bundles
+    - Stamp Observations with condition codes
+  - During the Deduplication Workflow, the Convert Step will:
+    - Generate a hash for the item
+    - Compare this hash against those batched in the same submitted report
+    - Compare this hash with others stored in the database
+    - Create appropriate action logs and Azure Events if any duplicates were detected
+  - After the Deduplication Workflow occurs, the Convert Step will perform necessary pipeline functions:
+    - Apply sender schema transform
+    - Make a report object with item lineage
+      - **This will now include the newly generated (or null) hash**
+    - Create a route event
+    - Upload to blobstore
+    - Update the database
+    - Add the report to the route destination filter step queue
+
+### Other Updates
+- Sender Settings: The Universal Pipeline will use the existing `allowDuplicates` sender setting. 
+  - By default, the deduplication workflow is enabled for UP senders (`allowDuplicates` = false). Individual receivers will set `allowDuplicates` to `true` to opt out.
+  - The abstract Sender class [should not initialize](https://github.com/CDCgov/prime-reportstream/blob/f48d719b876859169deb0360487f63965d8be5a0/prime-router/src/main/kotlin/Sender.kt#L60) `allowDuplicates`.
+  - The UniversalPipelineSender class [should default to false](https://github.com/CDCgov/prime-reportstream/blob/f48d719b876859169deb0360487f63965d8be5a0/prime-router/src/main/kotlin/Sender.kt#L196).
+- Receiver Step: There is stubbed code to be removed here: The SumissionReceiver refers to the once theorized area where UP deduplication would be invoked. ([SubmissionReceiver.kt#L284-L292](https://github.com/CDCgov/prime-reportstream/blob/0c5e0b058e35e09786942f2c8b41c1d67a5b1d16/prime-router/src/main/kotlin/SubmissionReceiver.kt#L284-L292))
+- Database Changes: Several changes will need to take place to improve efficiency. **Over 90% of the values stored in item_lineage.item_hash are not relevant to either CP or UP deduplication.**
+  - Immediate change: Modify item_lineage.item_hash to accept NULL values.
+    - Reasoning: 90%+ of item_lineage table rows have no need to store the item_hash. Allowing null values will reduce updates to the index and indexing times for the table. 
```

Comment:
Hmm, I wouldn't say they have NO reason to store these. They may not be relevant to deduplication but they do technically record the item in the DB long after it has been deleted from the blob store. This MAY be useful for debugging purposes but probably not... For example, someone says "I received this item from RS and its wrong" we could hash it and check against the DB if it is indeed what we sent or if it has been since modified

Reply:
That's a fair point but I think this will have a positive effect on performance around this table, especially long term as more null values are added. I think pulling the report from blob storage would be better for this kind of scenario, but I'm not sure how long those stick around for.